libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -2349,6 +2534,148 @@ IotHttpsReturnCode_t IotHttpsClient_SendSync(IotHttpsConnectionHandle_t connHand
 
 /*-----------------------------------------------------------*/
 
+IotHttpsReturnCode_t IotHttpsClient_WriteRequestBody(IotHttpsRequestHandle_t reqHandle, char *pBuf, uint32_t len, int isComplete)
+{
+    HTTPS_FUNCTION_ENTRY( IOT_HTTPS_OK );
+
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP(reqHandle);
+    /* This function is not valid for a synchronous response. Applications need to configure the request body in 
+       IotHttpsRequestInfo_t.pSyncInfo_t.reqData before calling IotHttpsClient_SendSync(). */
+    HTTPS_ON_ARG_ERROR_GOTO_CLEANUP(reqHandle->isAsync);
+    HTTPS_ON_ARG_ERROR_MSG_GOTO_CLEANUP(isComplete == 1,
+        IOT_HTTPS_NOT_SUPPORTED,
+        "isComplete must be 1 in IotHttpsClient_WriteRequestBody() for the current version of the HTTPS Client library.");
+
+    /* If the bodyLength is greater than 0, then we already called this function and we need to enforce that this
+       function must only be called once. We only call this function once so that we can calculate the Content-Length. */
+    if(reqHandle->bodyLength > 0)
+    {
+        IotLogError("Error this function must be called once with the data needed to send. Variable length HTTP "
+            "request body is not supported in this library.");
+        HTTPS_SET_AND_GOTO_CLEANUP(IOT_HTTPS_MESSAGE_FINISHED);
+    }
+
+    /* Set the pointer to the body and the length for the content-length calculation. */
+    reqHandle->pBody = pBuf;
+    reqHandle->bodyLength = len;
+
+    HTTPS_FUNCTION_EXIT_NO_CLEANUP();
+}
+
+/*-----------------------------------------------------------*/
+
+IotHttpsReturnCode_t IotHttpsClient_ReadResponseBody(IotHttpsResponseHandle_t respHandle, uint8_t * pBuf, uint32_t *pLen)
+{
+    HTTPS_FUNCTION_ENTRY( IOT_HTTPS_OK );
+
+    uint32_t bodyLengthInHeaderBuf = 0;
+
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP(respHandle);
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP(pBuf);
+    HTTPS_ON_NULL_ARG_GOTO_CLEANUP(pLen);
+    HTTPS_ON_ARG_ERROR_GOTO_CLEANUP(respHandle->isAsync);
+
+    /* Set the current body in the respHandle to use in _receiveHttpsBody(). _receiveHttpsBody is generic
+        to both async and sync request/response handling. In the sync version the body is configured during
+        initializing the request. In the async version the body is given in this function on the fly. */
+    respHandle->pBody = pBuf;
+    respHandle->pBodyCur = respHandle->pBody;
+    respHandle->pBodyEnd = respHandle->pBodyCur + *pLen;
```

Comment:
is it guaranteed that ```pBodyEnd``` is pointing within ```pBody``` ?

Reply:
In this case, it is up to the application to make sure that pLen is the size of it's buffer. There is no way for the library to verify this.